taiga_halo2/examples/taiga_sudoku/app_vp.rs

```diff
@@ -0,0 +1,800 @@
+                        layouter.namespace(|| "lhs * diff"),
+                        &cell_lhs,
+                        &diff,
+                    )
+                    .unwrap();
+                }
+            }
+            let cell_lhs_inv = assign_free_advice(
+                layouter.namespace(|| "non-zero sudoku_cell"),
+                config.advices[0],
+                cell_lhs.value().map(|x| x.invert().unwrap()),
+            )
+            .unwrap();
+
+            let cell_div = MulInstructions::mul(
+                &config.mul_chip(),
+                layouter.namespace(|| "lhs * 1/lhs"),
+                &cell_lhs,
+                &cell_lhs_inv,
+            )
+            .unwrap();
+
+            let constant_one = assign_free_constant(
+                layouter.namespace(|| "constant one"),
+                config.advices[0],
+                pallas::Base::one(),
+            )?;
+
+            layouter.assign_region(
+                || "lhs * 1/lhs = 1",
+                |mut region| region.constrain_equal(cell_div.cell(), constant_one.cell()),
+            )?;
+        }
+
+        Ok(())
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    fn check_state(
+        config: &SudokuStateCheckConfig,
+        mut layouter: impl Layouter<pallas::Base>,
+        is_spend_note: &AssignedCell<pallas::Base, pallas::Base>,
+        init_state: &AssignedCell<pallas::Base, pallas::Base>,
+        spend_note_pre_state: &AssignedCell<pallas::Base, pallas::Base>,
+        output_note_cur_state: &AssignedCell<pallas::Base, pallas::Base>,
+        spend_note_app_data_encode: &AssignedCell<pallas::Base, pallas::Base>,
+        spend_note: &SpendNoteVar,
+        output_note: &OutputNoteVar,
+    ) -> Result<(), Error> {
+        layouter.assign_region(
+            || "dealer intent check",
+            |mut region| {
+                config.assign_region(
+                    is_spend_note,
+                    init_state,
+                    &spend_note.app_data,
+                    spend_note_app_data_encode,
+                    &spend_note.app_vk,
+                    &output_note.app_vk,
+                    spend_note_pre_state,
+                    output_note_cur_state,
+                    0,
+                    &mut region,
+                )
+            },
+        )
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    fn check_solution(
+        mut layouter: impl Layouter<pallas::Base>,
+        state_update_config: &StateUpdateConfig,
+        triple_mul_config: &TripleMulConfig,
+        value_check_config: &ValueCheckConfig,
+        is_spend_note: &AssignedCell<pallas::Base, pallas::Base>,
+        pre_state: &[AssignedCell<pallas::Base, pallas::Base>],
+        cur_state: &[AssignedCell<pallas::Base, pallas::Base>],
+        _spend_note: &SpendNoteVar,
+        output_note: &OutputNoteVar,
+    ) -> Result<(), Error> {
+        // check state update: the cur_state is updated from pre_state
```

Comment:
I can't see how we are checking the current state is updated from the previous state. `state_update_config.assign_region` is just a few copy constraints

Reply:
There is a gate to check state updates: https://github.com/anoma/taiga/blob/8d7c6c975e12128478b48431b5f468c970e126d2/taiga_halo2/examples/taiga_sudoku/gadgets/state_update.rs#L55
It checks that the pre_state_cell is zero or pre_state_cell is equal with cur_state_cell